File: Python/day-2/day_2_part_2.py
def check_valid(arr):
    return (sorted(arr) == arr or sorted(arr, reverse=True) == arr) \
        and adjacent_is_near(arr)


def adjacent_is_near(arr):
    for i in range(len(arr) - 1):
        if not 1 <= abs(arr[i] - arr[i + 1]) <= 3:
            return False
    return True


def main():
    total = 0

    with open("input") as f:
        for line in f.readlines():
            # Turn line into array of integers
            splt = [int(x) for x in line.split()]
            if check_valid(splt):
                total += 1

            # If not valid, check if any permutation satisfies
            else:
                for i in range(len(splt)):
                    temp = splt[0:i] + splt[i + 1:]
                    if check_valid(temp):
                        total += 1
                        break
        print(total)


if __name__ == '__main__':
    main()

File: Java/day-2/src/Day2Part2.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.*;

public class Day2Part2 {
    // Check if every neighbour is between 1 and 3 difference
    public static boolean adjacentIsNear(ArrayList<Integer> a) {
        for (int i = 1; i < a.size(); i++) {
            if (!(Math.abs(a.get(i) - a.get(i - 1)) >= 1 && Math.abs(a.get(i) - a.get(i - 1)) <= 3 )) {
                return false;
            }
        }
        return true;
    }


    public static boolean isSortedAscending(ArrayList<Integer> a) {
        for (int i = 1; i < a.size(); i++) {
            if (a.get(i) > a.get(i - 1)) {
                return false;
            }
        }
        return true;
    }


    public static boolean isSortedDescending(ArrayList<Integer> a) {
        for (int i = 1; i < a.size(); i++) {
            if (a.get(i) < a.get(i - 1)) {
                return false;
            }
        }
        return true;
    }


    public static boolean isValid(ArrayList<Integer> a) {
        return (isSortedAscending(a) || isSortedDescending(a)) && adjacentIsNear(a);
    }


    public static void main(String[] args) {
        int total = 0;
        int currentInt;
        ArrayList<Integer> list = new ArrayList<>();

        File file = new File("src/input");
        try {
            Scanner sc = new Scanner(file);
            // Read each number into an ArrayList
            while (sc.hasNext()) {
                StringTokenizer st = new StringTokenizer(sc.nextLine());
                while (st.hasMoreTokens()) {
                    currentInt = Integer.parseInt(st.nextToken());
                    list.add(currentInt);
                    //System.out.printf(currentInt + " ");
                }
                if (isValid(list)) {
                    System.out.println("Sorted!: " + list.toString());
                    total += 1;
                } else {
                    // If not valid, check all possible permutations which satisfies
                    for (int i = 0; i < list.size(); i++) {
                        ArrayList<Integer> temp = new ArrayList<>(list);
                        temp.remove(i);
                        if (isValid(temp)) {
                            total += 1;
                            break;
                        }
                    }
                    System.out.println("Not sorted!: " + list.toString());
                }
                list.clear();
            }
            sc.close();
            System.out.println("Total: " + total);
        }
        catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        }
    }
}